// react-native-web/dist/index.js
'use strict';

exports.__esModule = true;
exports.PointPropType = exports.EdgeInsetsPropType = exports.ColorPropType = exports.VirtualizedList = exports.View = exports.TouchableWithoutFeedback = exports.TouchableOpacity = exports.TouchableNativeFeedback = exports.TouchableHighlight = exports.Touchable = exports.TextInput = exports.Text = exports.Switch = exports.StatusBar = exports.Slider = exports.SectionList = exports.ScrollView = exports.SafeAreaView = exports.RefreshControl = exports.ProgressBar = exports.Picker = exports.Modal = exports.ListView = exports.KeyboardAvoidingView = exports.ImageBackground = exports.Image = exports.FlatList = exports.CheckBox = exports.Button = exports.ART = exports.ActivityIndicator = exports.Vibration = exports.UIManager = exports.StyleSheet = exports.Platform = exports.PixelRatio = exports.PanResponder = exports.NetInfo = exports.Linking = exports.Keyboard = exports.InteractionManager = exports.I18nManager = exports.Easing = exports.Dimensions = exports.Clipboard = exports.BackHandler = exports.AsyncStorage = exports.AppState = exports.AppRegistry = exports.Animated = exports.ViewPropTypes = exports.TextPropTypes = exports.NativeModules = exports.processColor = exports.unmountComponentAtNode = exports.render = exports.findNodeHandle = exports.createElement = undefined;

var _createElement = require('./exports/createElement');

var _createElement2 = _interopRequireDefault(_createElement);

var _findNodeHandle = require('./exports/findNodeHandle');

var _findNodeHandle2 = _interopRequireDefault(_findNodeHandle);

var _processColor = require('./exports/processColor');

var _processColor2 = _interopRequireDefault(_processColor);

var _render = require('./exports/render');

var _render2 = _interopRequireDefault(_render);

var _unmountComponentAtNode = require('./exports/unmountComponentAtNode');

var _unmountComponentAtNode2 = _interopRequireDefault(_unmountComponentAtNode);

var _NativeModules = require('./exports/NativeModules');

var _NativeModules2 = _interopRequireDefault(_NativeModules);

var _TextPropTypes = require('./exports/TextPropTypes');

var _TextPropTypes2 = _interopRequireDefault(_TextPropTypes);

var _ViewPropTypes = require('./exports/ViewPropTypes');

var _ViewPropTypes2 = _interopRequireDefault(_ViewPropTypes);

var _Animated = require('./exports/Animated');

var _Animated2 = _interopRequireDefault(_Animated);

var _AppRegistry = require('./exports/AppRegistry');

var _AppRegistry2 = _interopRequireDefault(_AppRegistry);

var _AppState = require('./exports/AppState');

var _AppState2 = _interopRequireDefault(_AppState);

var _AsyncStorage = require('./exports/AsyncStorage');

var _AsyncStorage2 = _interopRequireDefault(_AsyncStorage);

var _BackHandler = require('./exports/BackHandler');

var _BackHandler2 = _interopRequireDefault(_BackHandler);

var _Clipboard = require('./exports/Clipboard');

var _Clipboard2 = _interopRequireDefault(_Clipboard);

var _Dimensions = require('./exports/Dimensions');

var _Dimensions2 = _interopRequireDefault(_Dimensions);

var _Easing = require('./exports/Easing');

var _Easing2 = _interopRequireDefault(_Easing);

var _I18nManager = require('./exports/I18nManager');

var _I18nManager2 = _interopRequireDefault(_I18nManager);

var _Keyboard = require('./exports/Keyboard');

var _Keyboard2 = _interopRequireDefault(_Keyboard);

var _InteractionManager = require('./exports/InteractionManager');

var _InteractionManager2 = _interopRequireDefault(_InteractionManager);

var _Linking = require('./exports/Linking');

var _Linking2 = _interopRequireDefault(_Linking);

var _NetInfo = require('./exports/NetInfo');

var _NetInfo2 = _interopRequireDefault(_NetInfo);

var _PanResponder = require('./exports/PanResponder');

var _PanResponder2 = _interopRequireDefault(_PanResponder);

var _PixelRatio = require('./exports/PixelRatio');

var _PixelRatio2 = _interopRequireDefault(_PixelRatio);

var _Platform = require('./exports/Platform');

var _Platform2 = _interopRequireDefault(_Platform);

var _StyleSheet = require('./exports/StyleSheet');

var _StyleSheet2 = _interopRequireDefault(_StyleSheet);

var _UIManager = require('./exports/UIManager');

var _UIManager2 = _interopRequireDefault(_UIManager);

var _Vibration = require('./exports/Vibration');

var _Vibration2 = _interopRequireDefault(_Vibration);

var _ActivityIndicator = require('./exports/ActivityIndicator');

var _ActivityIndicator2 = _interopRequireDefault(_ActivityIndicator);

var _ART = require('./exports/ART');

var _ART2 = _interopRequireDefault(_ART);

var _Button = require('./exports/Button');

var _Button2 = _interopRequireDefault(_Button);

var _CheckBox = require('./exports/CheckBox');

var _CheckBox2 = _interopRequireDefault(_CheckBox);

var _FlatList = require('./exports/FlatList');

var _FlatList2 = _interopRequireDefault(_FlatList);

var _Image = require('./exports/Image');

var _Image2 = _interopRequireDefault(_Image);

var _ImageBackground = require('./exports/ImageBackground');

var _ImageBackground2 = _interopRequireDefault(_ImageBackground);

var _KeyboardAvoidingView = require('./exports/KeyboardAvoidingView');

var _KeyboardAvoidingView2 = _interopRequireDefault(_KeyboardAvoidingView);

var _ListView = require('./exports/ListView');

var _ListView2 = _interopRequireDefault(_ListView);

var _Modal = require('./exports/Modal');

var _Modal2 = _interopRequireDefault(_Modal);

var _Picker = require('./exports/Picker');

var _Picker2 = _interopRequireDefault(_Picker);

var _ProgressBar = require('./exports/ProgressBar');

var _ProgressBar2 = _interopRequireDefault(_ProgressBar);

var _RefreshControl = require('./exports/RefreshControl');

var _RefreshControl2 = _interopRequireDefault(_RefreshControl);

var _SafeAreaView = require('./exports/SafeAreaView');

var _SafeAreaView2 = _interopRequireDefault(_SafeAreaView);

var _ScrollView = require('./exports/ScrollView');

var _ScrollView2 = _interopRequireDefault(_ScrollView);

var _SectionList = require('./exports/SectionList');

var _SectionList2 = _interopRequireDefault(_SectionList);

var _Slider = require('./exports/Slider');

var _Slider2 = _interopRequireDefault(_Slider);

var _StatusBar = require('./exports/StatusBar');

var _StatusBar2 = _interopRequireDefault(_StatusBar);

var _Switch = require('./exports/Switch');

var _Switch2 = _interopRequireDefault(_Switch);

var _Text = require('./exports/Text');

var _Text2 = _interopRequireDefault(_Text);

var _TextInput = require('./exports/TextInput');

var _TextInput2 = _interopRequireDefault(_TextInput);

var _Touchable = require('./exports/Touchable');

var _Touchable2 = _interopRequireDefault(_Touchable);

var _TouchableHighlight = require('./exports/TouchableHighlight');

var _TouchableHighlight2 = _interopRequireDefault(_TouchableHighlight);

var _TouchableNativeFeedback = require('./exports/TouchableNativeFeedback');

var _TouchableNativeFeedback2 = _interopRequireDefault(_TouchableNativeFeedback);

var _TouchableOpacity = require('./exports/TouchableOpacity');

var _TouchableOpacity2 = _interopRequireDefault(_TouchableOpacity);

var _TouchableWithoutFeedback = require('./exports/TouchableWithoutFeedback');

var _TouchableWithoutFeedback2 = _interopRequireDefault(_TouchableWithoutFeedback);

var _View = require('./exports/View');

var _View2 = _interopRequireDefault(_View);

var _VirtualizedList = require('./exports/VirtualizedList');

var _VirtualizedList2 = _interopRequireDefault(_VirtualizedList);

var _ColorPropType = require('./exports/ColorPropType');

var _ColorPropType2 = _interopRequireDefault(_ColorPropType);

var _EdgeInsetsPropType = require('./exports/EdgeInsetsPropType');

var _EdgeInsetsPropType2 = _interopRequireDefault(_EdgeInsetsPropType);

var _PointPropType = require('./exports/PointPropType');

var _PointPropType2 = _interopRequireDefault(_PointPropType);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// APIs
exports.createElement = _createElement2.default;
exports.findNodeHandle = _findNodeHandle2.default;
exports.render = _render2.default;
exports.unmountComponentAtNode = _unmountComponentAtNode2.default;
exports.processColor = _processColor2.default;
exports.NativeModules = _NativeModules2.default;
exports.TextPropTypes = _TextPropTypes2.default;
exports.ViewPropTypes = _ViewPropTypes2.default;
exports.Animated = _Animated2.default;
exports.AppRegistry = _AppRegistry2.default;
exports.AppState = _AppState2.default;
exports.AsyncStorage = _AsyncStorage2.default;
exports.BackHandler = _BackHandler2.default;
exports.Clipboard = _Clipboard2.default;
exports.Dimensions = _Dimensions2.default;
exports.Easing = _Easing2.default;
exports.I18nManager = _I18nManager2.default;
exports.InteractionManager = _InteractionManager2.default;
exports.Keyboard = _Keyboard2.default;
exports.Linking = _Linking2.default;
exports.NetInfo = _NetInfo2.default;
exports.PanResponder = _PanResponder2.default;
exports.PixelRatio = _PixelRatio2.default;
exports.Platform = _Platform2.default;
exports.StyleSheet = _StyleSheet2.default;
exports.UIManager = _UIManager2.default;
exports.Vibration = _Vibration2.default;
exports.ActivityIndicator = _ActivityIndicator2.default;
exports.ART = _ART2.default;
exports.Button = _Button2.default;
exports.CheckBox = _CheckBox2.default;
exports.FlatList = _FlatList2.default;
exports.Image = _Image2.default;
exports.ImageBackground = _ImageBackground2.default;
exports.KeyboardAvoidingView = _KeyboardAvoidingView2.default;
exports.ListView = _ListView2.default;
exports.Modal = _Modal2.default;
exports.Picker = _Picker2.default;
exports.ProgressBar = _ProgressBar2.default;
exports.RefreshControl = _RefreshControl2.default;
exports.SafeAreaView = _SafeAreaView2.default;
exports.ScrollView = _ScrollView2.default;
exports.SectionList = _SectionList2.default;
exports.Slider = _Slider2.default;
exports.StatusBar = _StatusBar2.default;
exports.Switch = _Switch2.default;
exports.Text = _Text2.default;
exports.TextInput = _TextInput2.default;
exports.Touchable = _Touchable2.default;
exports.TouchableHighlight = _TouchableHighlight2.default;
exports.TouchableNativeFeedback = _TouchableNativeFeedback2.default;
exports.TouchableOpacity = _TouchableOpacity2.default;
exports.TouchableWithoutFeedback = _TouchableWithoutFeedback2.default;
exports.View = _View2.default;
exports.VirtualizedList = _VirtualizedList2.default;
exports.ColorPropType = _ColorPropType2.default;
exports.EdgeInsetsPropType = _EdgeInsetsPropType2.default;
exports.PointPropType = _PointPropType2.default;

// propTypes


// components

var ReactNative = {
  // top-level API
  createElement: _createElement2.default,
  findNodeHandle: _findNodeHandle2.default,
  render: _render2.default,
  unmountComponentAtNode: _unmountComponentAtNode2.default,
  // modules
  processColor: _processColor2.default,
  NativeModules: _NativeModules2.default,
  TextPropTypes: _TextPropTypes2.default,
  ViewPropTypes: _ViewPropTypes2.default,
  // APIs
  Animated: _Animated2.default,
  AppRegistry: _AppRegistry2.default,
  AppState: _AppState2.default,
  AsyncStorage: _AsyncStorage2.default,
  BackHandler: _BackHandler2.default,
  Clipboard: _Clipboard2.default,
  Dimensions: _Dimensions2.default,
  Easing: _Easing2.default,
  I18nManager: _I18nManager2.default,
  InteractionManager: _InteractionManager2.default,
  Keyboard: _Keyboard2.default,
  Linking: _Linking2.default,
  NetInfo: _NetInfo2.default,
  PanResponder: _PanResponder2.default,
  PixelRatio: _PixelRatio2.default,
  Platform: _Platform2.default,
  StyleSheet: _StyleSheet2.default,
  UIManager: _UIManager2.default,
  Vibration: _Vibration2.default,
  // components
  ActivityIndicator: _ActivityIndicator2.default,
  ART: _ART2.default,
  Button: _Button2.default,
  CheckBox: _CheckBox2.default,
  FlatList: _FlatList2.default,
  Image: _Image2.default,
  ImageBackground: _ImageBackground2.default,
  KeyboardAvoidingView: _KeyboardAvoidingView2.default,
  ListView: _ListView2.default,
  Modal: _Modal2.default,
  Picker: _Picker2.default,
  ProgressBar: _ProgressBar2.default,
  RefreshControl: _RefreshControl2.default,
  SafeAreaView: _SafeAreaView2.default,
  ScrollView: _ScrollView2.default,
  SectionList: _SectionList2.default,
  Slider: _Slider2.default,
  StatusBar: _StatusBar2.default,
  Switch: _Switch2.default,
  Text: _Text2.default,
  TextInput: _TextInput2.default,
  Touchable: _Touchable2.default,
  TouchableHighlight: _TouchableHighlight2.default,
  TouchableNativeFeedback: _TouchableNativeFeedback2.default,
  TouchableOpacity: _TouchableOpacity2.default,
  TouchableWithoutFeedback: _TouchableWithoutFeedback2.default,
  View: _View2.default,
  VirtualizedList: _VirtualizedList2.default,
  // propTypes
  ColorPropType: _ColorPropType2.default,
  EdgeInsetsPropType: _EdgeInsetsPropType2.default,
  PointPropType: _PointPropType2.default
};

exports.default = ReactNative;
